### PR TITLE
First fix for not immune to while(1){}

### PR DESCRIPTION
**Pull request by XmiliaH**
This does not fix the Promise & async problems

### CODE REVIEW

**Comment by XmiliaH** [comment]
And for the async problem I currently have no idea to fix it.

File: lib/main.js [hunk 1]
```diff
@@ -159,9 +173,25 @@ function loadScript(filename) {
 const SCRIPT_CACHE = {
 	cf: loadScript(`${__dirname}/contextify.js`).wrap('(function(require, host) { ', '\n})')._compileVM(),
 	sb: loadScript(`${__dirname}/sandbox.js`).wrap('(function (vm, host, Contextify, Decontextify, Buffer) { ', '\n})')._compileVM(),
-	exp: new VMScript('({exports: {}})')._compileVM()
+	exp: new VMScript('({exports: {}})')._compileVM(),
+	runTimeout: new VMScript('fn()', 'timeout_bridge.js')._compileVM()
 };
 
+const TIMEOUT_CONTEXT = vm.createContext();
```
Comment (patriksimek) [inline lib/main.js]: I'd suggest changing that to `const TIMEOUT_CONTEXT = null` and moving `vm.createContext()` inside the `doWithTimeout` function. Just a performance tweak for situations when timeout is not used.

File: lib/main.js [hunk 2]
```diff
@@ -292,15 +319,13 @@ class VM extends EventEmitter {
...
-		try {
-			return this._internal.Decontextify.value(script._compiledVM.runInContext(this._context, {
-				filename: script.filename,
-				displayErrors: false,
-				timeout: this.options.timeout
-			}));
-		} catch (e) {
-			throw this._internal.Decontextify.value(e);
-		}
+		return doWithTimeout(()=>{
```
Comment (patriksimek) [inline lib/main.js]: We should use `doWithTimeout` only when timeout is specified. Related to my previous comment.

**Comment by patriksimek** [comment]
Thanks.

**Comment by XmiliaH** [comment]
@OrkhanAlikhanov see https://github.com/patriksimek/vm2/issues/180#issuecomment-531472263. However, there is currently a PR https://github.com/patriksimek/vm2/pull/246 that tries to address this issue by disallowing scripts with the word async and patching Promise. This has some sideeffects, for example eval can't be used as direct eval any more.

**Comment by OrkhanAlikhanov** [comment]
Thank you for the quick response!

**Comment by fangk** [comment]
https://medium.com/@bvjebin/js-infinite-loops-killing-em-e1c2f5f2db7f
can we use babel inject code into WhileStatement and ForStatement?

**Pull request merged**